File: LinkedLists/circularDoublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <ctype.h>

typedef struct Node
{
    int data;
    struct Node *next;
    struct Node *pvs;
} node;

node *head = NULL;
node *tail = NULL;

void insertAtHead(int data);
void deleteAtHead();
void insertAtTail(int data);
void deleteAtTail();
void insertAtPos(int data, int pos);
void deleteAtPos(int pos);
void freeAllNodes();
void freeNode(node *nodeToFree);
int getLengthOfCircularDoublyLinkedList();
void displayCircularDoublyLinkedList();
void displayReverseCircularDoublyLinkedList();

int main()
{
    insertAtHead(10000);
    // insertAtTail(99);
    insertAtHead(30000);
    insertAtHead(5000);
    insertAtHead(2000);
    // insertAtTail(100);
    // insertAtPos(1,7);
    // deleteAtHead();
    // deleteAtHead();
    // deleteAtTail();
    // deleteAtTail();
    deleteAtPos(5);
    // reverseDoublyLinkedList();
    // bubbleSortDoublyLinkedList();
    displayCircularDoublyLinkedList();
    displayReverseCircularDoublyLinkedList();

    return 0;
}

void insertAtHead(int data)
{
    node *link = (node *)malloc(sizeof(node));

    if (link == NULL)
    {
        printf("Memory Allocation Failed\n");
        return;
    }

    link->data = data;

    if ((head == NULL) && (tail == NULL))
    {
        link->next = link;
        link->pvs = link;
        head = link;
        tail = link;

        return;
    }

    head->pvs = link;
    link->next = head;
    link->pvs = tail;
    tail->next = link;
    head = link;

    return;
}

void deleteAtHead()
{
    if ((head == NULL) && (tail == NULL))
    {
        printf("The CircularDoublyLinkedList is Empty\n");
        return;
    }

    if (head == tail)
    {
        freeNode(head);
        head = NULL;
        tail = NULL;

        return;
    }

    node *newHead = head->next;
    newHead->pvs = tail;
    tail->next = newHead;
    freeNode(head);
    head = newHead;

    return;
}

void insertAtTail(int data)
{
    node *link = (node *)malloc(sizeof(node));

    if (link == NULL)
    {
        printf("Memory Allocation Failed\n");
        return;
    }

    link->data = data;

    if ((head == NULL) && (tail == NULL))
    {
        head = link;
        link->pvs = link;
        link->next = link;
        tail = link;
        return;
    }

    link->pvs = tail;
    tail->next = link;
    link->next = head;
    head->pvs = link;
    tail = link;

    return;
}

void deleteAtTail()
{
    if ((head == NULL) && (tail == NULL))
    {
        printf("The Circular DoublyLinkedList is Empty\n");
        return;
    }

    if (head == tail)
    {
        freeNode(head);
        head = NULL;
        tail = NULL;
        return;
    }

    node *newTail = tail->pvs;
    newTail->next = head;
    head->pvs = newTail;
    freeNode(tail);
    tail = newTail;

    return;
}

void insertAtPos(int data, int pos)
{
    node *link = (node *)malloc(sizeof(node));

    if (link == NULL)
    {
        printf("Memory Allocation Failed\n");
        return;
    }

    if (pos == 1)
    {
        insertAtHead(data);
        return;
    }

    int curPos = 2;
    node *curNode = head->next;
    node *pvsNode = head;
    while ((curPos != pos) && (curNode != head))
    {
        curPos++;
        pvsNode = curNode;
        curNode = curNode->next;
    }

    link->data = data;
    pvsNode->next = link;
    link->pvs = pvsNode;
    link->next = curNode;

    if (curNode == head)
    {
        tail = link;
        head->pvs = link;

        return;
    }

    curNode->pvs = link;

    return;
}

void deleteAtPos(int pos)
{
    if ((head == NULL) && (tail == NULL))
    {
        printf("The SinglyLinkedList is empty\n");
        return;
    }

    if (pos == 1)
    {
        deleteAtHead();
        return;
    }

    int curPos = 2;
    node *curNode = head->next;
    node *pvsNode = head;

    while ((curPos != pos) && (curNode != head))
    {
        curPos++;
        pvsNode = curNode;
        curNode = curNode->next;
    }

    if (curNode == head)
    {
        printf("Invalid delete position\n");
        return;
    }

    pvsNode->next = curNode->next;
    (curNode->next)->pvs = pvsNode;
    tail = pvsNode;
    freeNode(curNode);

    return;
}

void freeAllNodes()
{
    node *curNode = head;
    while (curNode != NULL)
    {
        node *tempNode = curNode;
        curNode = curNode->next;

        freeNode(tempNode);
    }

    head = NULL;
    tail = NULL;
    return;
}

void freeNode(node *nodeToFree)
{
    memset(nodeToFree, 0, sizeof(node));
    free(nodeToFree);

    return;
}

int getLengthOfCircularDoublyLinkedList()
{
    int length = 0;
    node *curNode = head;

    while (curNode->next != head)
    {
        curNode = curNode->next;
        length++;
    }

    return length + 1;
}

void displayCircularDoublyLinkedList()
{
    if ((head == NULL) && (tail == NULL))
    {
        printf("Circular Doubly linkedlist is empty\n");
        return;
    }

    node *curNode = head;
    int pos = 1;

    while (curNode->next != head)
    {
        printf("The element as position %d is %d\n", pos, curNode->data);
        curNode = curNode->next;
        pos++;
    }
    printf("The element as position %d is %d\n", pos, curNode->data);
    return;
}

void displayReverseCircularDoublyLinkedList()
{
    if ((head == NULL) && (tail == NULL))
    {
        printf("Circular Doubly linkedlist is empty\n");
        return;
    }

    node *curNode = tail;

    while (curNode->pvs != tail)
    {
        printf("The element is %d\n", curNode->data);
        curNode = curNode->pvs;
    }

    printf("The element is %d\n", curNode->data);
    return;
}

File: PythonDSACodes/circularDoublyLinkedList.py
class node():
    def __init__(self,data,prev=None,next=None):
        self.data=data
        self.next=next
        self.prev=prev
    
class circularDoublyLinkedList():
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_at_beginning(self,data):

        newNode = node(data,None,self.head)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
            self.head.next = newNode
            self.head.prev = newNode

        elif self.head != None:
            self.head.prev = newNode
            self.head = newNode
            self.head.prev = self.tail
            self.tail.next = newNode

    
    def printCircularDoublyLinkedList(self):
        if self.head == None:
            print("Doubly Linked List is Empty")
            return

        itr = self.head
        doublyLinkedListStr = "" 
        while itr:
            doublyLinkedListStr += str(itr.data) + " ↔ "
            if itr.next == self.head:
                break
            else:
                itr = itr.next
        
        print(doublyLinkedListStr)

    def insert_at_end(self,data):
        if self.head == None or self.tail == None:
            self.head = node(data,None,None)
            self.tail = self.head
            self.head.prev = self.head
            self.head.next = self.head
            return

        elif self.head != None:
            itr = self.head
            while itr:
                if itr.next == self.head:
                    newnode = node(data,self.tail,self.head)
                    itr.next = newnode
                    self.head.prev = newnode
                    self.tail = newnode
                    break
                itr = itr.next
    
    def printCircularDoublyLinkedListReverse(self):
        if self.tail == None:
            print("Doubly Linked List is Empty")
            return

        itr = self.tail
        doublyLinkedListStr = "" 
        while itr:
            doublyLinkedListStr += str(itr.data) + " ↔ "
            if itr.prev == self.tail:
                break
            else:
                itr = itr.prev
        
        print(doublyLinkedListStr)

l = circularDoublyLinkedList()
l.insert_at_end(9)
l.insert_at_end(2)
# l.insert_at_beginning(3)
# l.insert_at_beginning(4)
l.insert_at_end(10)
# l.insert_at_end(11)
# l.insert_at_end(12)
# l.insert_at_beginning(5)


l.printCircularDoublyLinkedList()
print(l.head.data)
print(l.head.next.data)
print(l.head.next.next.data)
# print(l.head.next.next.next.data)
# print(l.head.next.next.next.next.data)
# print(l.head.next.next.next.next.next.data)
l.printCircularDoublyLinkedListReverse()
print(l.tail.data)
print(l.tail.prev.data)
print(l.tail.prev.prev.data)
print(l.tail.prev.prev.prev.data)
print(l.tail.prev.prev.prev.prev.data)
print(l.tail.prev.prev.prev.prev.prev.data)
